### ISSUE TITLE
Xtensive.Orm 7.2beta-1 Authentication issue

### ISSUE THREAD

**Issue by AnamikaDe8** [DataObjects-NET/dataobjects-net#393]
@SergeiPavlov @alexyakunin @shuruev @AlexUstinov 
We have a question.
 Background

A .Net 4.7.2 web project with razor pages has been upgraded to .net 8.
1.	Servicestack 3 upgraded to ServiceStack 8
2.	Xtensive.Orm 5  7.2-Beta

We are facing an issue while trying to authenticate “auth/credentials” using IAuthenticationService Interface of Xtensive.orm.Security.

The IAuthenticationService.Authenticate returns null error

protected override IPrincipal Authenticate(string userNameOrEmail, string password)
{
     var config = this.Session.GetSecurityConfiguration();
     var service = this.Session.Services.Get<IServiceStackHashingService>(config.HashingServiceName);Error here 

     if (service == null)
         throw new InvalidOperationException(string.Format("Hashing service by name {0} is not found. Check Xtensive.Security configuration", config.HashingServiceName));
}
![error](https://github.com/user-attachments/assets/a17bd5b3-6955-4839-ae62-3ffdd186ac29)
HashingServiceName = plain
 

We are guessing, the services are not getting registered properly.

We have used Domain Factory to initialize the DomainConfig and implemented a auth class from IAuthenticationService.

Could you please help.

**Comment by SergeiPavlov** [comment]
+ @alex-kulakov

**Comment by alex-kulakov** [comment]
Hello guys, thanks Sergei for pointing me to this.

@AnamikaDe8,

Let's check domain configuration types first. You need to make sure that types from  ```Xtensive.Orm.Security``` assembly registered in ```DomainConfiguration.Types``` collection. If you have your own implementations of some interfaces from ```Xtensive.Orm.Security``` assembly then they should be registered too.  You can actually see the types which are registered while debugging if you check ```DomainConfiguration``` instance and its property ```Types``` (you may have to click ResultView).

Please, check the types and tell me whether there are some types from ```Xtensive.Orm.Security``` assembly or maybe something missing that should be in registered types?

All session services (the ones in session.Services) are instantiated from types that are registered in domain configuration, if there is no type given in ```session.Services.Get<T>(seviceName)``` it will return null. This is why I asked to check the types



Another thing that bothers me is that "plain" hashing service appeared. If you don't use it and this happens then we probably have another issue. Plain hashing provider is default value. It can indicate that configuration section wasn't read from configuration file.

I don't know whether you are aware so I'll try to explain one particularity. Here's the thing,
```.Net Framework``` and ```.NET``` (don't confuse them) have different approach to ```*.config``` files. ```.Net Framework``` kind of merges config files of all assemblies loaded into ```AppDomain``` into one single config and you can access sections of it by doing ConfigurationManager.GetSection(someSectionName) and no matter in what file the section was it will be there for the manager.

.NET don't act like that (at least last time I checked). It gets only config file of running executable. For example, NUnit runner which runs tests from some project will not have access to the project's App.config file. We faced this issue and had to load configurations explicitly in our unit-test projects. For example,

```csharp
    private System.Configuration.Configuration GetConfigurationForTestAssembly()
    {
      // this gets true assembly configuration instead of config of nunit runner
      return GetType().Assembly.GetAssemblyConfiguration();
    }
```
and later on, load our configurations from the given instance

```csharp
    protected DomainConfiguration LoadDomainConfiguration(string sectionName, string name)
    {
      return DomainConfiguration.Load(GetConfigurationForTestAssembly(), sectionName, name);
    }
```

I'm telling you this because it might be a part of problem and it's good to explain the root that causes it.

Thanks god, it is easy to workaround - you just need to load ```SecurityConfiguration``` and put it to ```Domain.Extensions``` collection after domain is built. For example, here is an extension method which may help you.

```csharp
    public static SecurityConfiguration GetSecurityConfiguration(this Domain domain, System.Configuration.Configuration configuration)
    {
      var result = domain.Extensions.Get<SecurityConfiguration>();
      if (result == null) {
        result = SecurityConfiguration.Load(configuration);
        domain.Extensions.Set(result);
      }
      return result;
    }
```

I will probably put something similar to the ```Xtensive.Orm.Security``` project.

**Comment by AnamikaDe8** [comment]
Thanks @alex-kulakov For the reply. 
Please find the findings
The Types screenshot given below
![image](https://github.com/user-attachments/assets/cfbb2226-488f-4346-8328-39e36981ced7)

**Comment by alex-kulakov** [comment]
So, the issue is inability to read config.

I'm not sure about how to correctly convert web.config to json config or how to connect old one it to asp.net application. Probably, I will have to re-think configuration loading in projects. I will consult with colleagues from web development department to not give you wrong answer, but unfortunately it will happen only on Monday.

If it blocks development process, I can suggest to temporary define domain and security configurations in code instead of configuration file.

For ```SecurityConfuguration```, just create an instance and put it to the Extensions collection like so:

```csharp
    public void InitializeSecurity(this Domain domain)
    {
        var secConfiguration = new SecurityConfiguration() {
            HashingServiceName = "servicestack",
            AuthenticationServiceName = "servicestack"
        };
        domain.Extensions.Set(secConfiguration);
    }
```

Do you need help with DomainConfiguration? Most of properties of it are self-explanatory, maybe ```DomainConfiguration.Types``` collection can confuse you, you need to use ```Register()``` method, not ```Add()```, it has variant which gets ```Assembly``` as parameter.

**Comment by AnamikaDe8** [comment]
Hi, We tried
    public void InitializeSecurity(this Domain domain)
    {
        var secConfiguration = new SecurityConfiguration() {
            HashingServiceName = "servicestack",
            AuthenticationServiceName = "servicestack"
        };
        domain.Extensions.Set(secConfiguration);
    }

But it seems the hashingservice and AuthenticationServiceName both are private Set. Hence cannot be set.

**Comment by AnamikaDe8** [comment]
hi @alex-kulakov ,

We have sucessfully implemented the ORM.Config loading of configuration. Earlier we loaded the Auth and Hash Service programatically
![image](https://github.com/user-attachments/assets/ca57673a-37b1-49aa-91c0-dd3f657215d8)

Though, 7.2-Beta-1 being the only version which included the .NET 8 support, we cannot move our code with a Beta version. Could we expect A final release with .Net 8 support any time soon.

it would be great, if the Configuration loading from appsettings.json implementation also is included.

Thanks in advance

**Comment by alex-kulakov** [comment]
@AnamikaDe8,

I'm working on the loading of configurations form IConfigurationSection, it will support both json and xml (ConfigurationManager compatible format and newer version without xml attributes). Algorithms for extensions are already implemented, so I'm making proper testing and "polishing". Domain configuration, though, is different story, it is complicated so it will take some time - couple of weeks at least (if less time consumed, I will be happy).

Speaking of release, I need to publish  7.1.2 first, then there are pull-requests I plan to merge to master and, since some of them contain breaking changes and probably some issues may appear, next version from master will be 7.2-Beta-2, if everything is ok - RC  with slight changes, if RC is OK then finally release version. Usually, I'd like pre-release versions to stay published for several weeks before next version to let possible issues pop. We definitely make it released 😉

Honestly, I had to make some bug fixes and other changes in our previous releases, some very important customers still use 6.0 so I have to make it live. This is why I was away from the version in master but after 7.1.2 will be releases all my attention will be on master version.

**Comment by alex-kulakov** [comment]
@AnamikaDe8, I just created PR #397 which should make cases like yours a bit easier, at least no workaround will be required. after 7.1.2 release the changes will be merged to master and later released as next 7.2 version.

I believe I can close this issue now, can't I?
